Utils/ObjectPool.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace DUCK.Pooling
+{
+	public static class ObjectPool
+	{
+		private interface Pool
+		{
+			Type ObjectType { get; }
+
+			void Add(PoolableObject obj);
+		}
+
+		private class Pool<T> : Pool 
+			where T : PoolableObject
+		{
+			private Queue<T> pooledObjects = new Queue<T>();
+
+			public Type ObjectType
+			{
+				get
+				{
+					return typeof(T);
+				}
+			}
+
+			public T Get()
+			{
+				var obj = (pooledObjects.Count > 0)
+					? pooledObjects.Dequeue()
+					: null;
+
+				if (obj)
+				{
+					obj.RemoveFromPool();
+				}
+
+				return obj;
+			}
+
+			public void Add(PoolableObject obj)
+			{
+				if (obj != null)
+				{
+					obj.AddToPool();
+					pooledObjects.Enqueue((T)obj);
+				}
+			}
+		}
+
+		private static Dictionary<Type, Pool> poolLookup = new Dictionary<Type, Pool>();
+
+		public static T Instantiate<T>(T original = null) where T : PoolableObject
+		{
+			T obj = null;
+
+			if (PoolExists(typeof(T)))
+			{
+				obj = GetFromPool<T>();
+
+				if (obj != null)
+				{
+					if (original != null)
+					{
+						obj.CopyFrom(original);
+					}
+
+					return obj;
+				}
+			}
+			else
+			{
+				CreatePool<T>();
+			}
+
+			return obj ?? GameObject.Instantiate<T>(original);
+		}
+
+		public static T Instantiate<T>(string resourcePath) where T : PoolableObject
+		{
+			T obj = null;
+
+			if (PoolExists(typeof(T)))
+			{
+				obj = GetFromPool<T>();
+			}
+			else
+			{
+				CreatePool<T>();
+			}
+
+			return obj ?? Resources.Load<T>(resourcePath);
+		}
+
+		public static void Destroy(PoolableObject obj)
```

Comment:
When an object is released to a pool it would be good to invoke a cleanup step.

Reply:
What would you be cleaning up? PoolableObject already expoded AddedToPool and RemovedFromPool events for any components on the gameObject to listen for - this is where I'd expect custom cleanup actions to take place

Reply:
I'd be cleaning up events :) and state. I didn't spot those add and remove methods though.

Not sure if publicly subscribe-able events will eventually prove problematic in a system that is dealing with objects that at any given time can be hard or soft destroyed.